### PR TITLE
feat(core): add v1 feedback tools and structured report storage

### LINKED ISSUE #399: feat: AOC Li feedback tools for bug_report and improvement_suggestion (v1)
## Scope\nAdd structured feedback tools for decentralized self-improvement.\n\n### Requirements\n- Add tool contracts/schemas for bug_report and improvement_suggestion\n- Add validation (required fields: title/body/evidence/session/source/confidence)\n- Persist reports in structured storage\n- Add simple dedupe key generation\n\n### Constraints\n- Do not implement fork ensure/bootstrap here\n- Do not implement report->issue automation here\n\n### Validation\n- Add/adjust schema + storage tests\n- Include examples in docs

### PR DESCRIPTION
**Pull request by AgentWrapper**
## Summary
- add explicit v1 tool contracts/schemas for `bug_report` and `improvement_suggestion`
- enforce required fields (`title`, `body`, `evidence`, `session`, `source`, `confidence`) with strict validation
- add `FeedbackReportStore` with structured key=value persistence under `feedback-reports/`
- add deterministic dedupe key generation (`sha256` over normalized payload)
- expose new API from `@composio/ao-core` and document usage/examples (migration note included: no migration required)

## Scope Guardrails
- does **not** implement fork ensure/bootstrap
- does **not** implement report -> issue automation

## Tests
- `pnpm --filter @composio/ao-core typecheck`
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts src/__tests__/paths.test.ts`

## Issue
Closes #399

### CODE REVIEW

File: packages/core/src/feedback-tools.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    const trimmed = line.trim();
+    if (!trimmed || trimmed.startsWith("#")) continue;
+    const eqIndex = trimmed.indexOf("=");
+    if (eqIndex === -1) continue;
+    const key = trimmed.slice(0, eqIndex).trim();
+    const value = trimmed.slice(eqIndex + 1).trim();
+    if (key) result[key] = value;
+  }
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply packages/core/src/feedback-tools.ts]: Addressed in commit 28f2d6a.

Resolution:
- Removed duplicated key/value parser logic by extracting shared helper `parseKeyValueContent` in `packages/core/src/key-value.ts`.
- Reused it from both `packages/core/src/metadata.ts` and `packages/core/src/feedback-tools.ts`.

Evidence:
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts src/__tests__/metadata.test.ts` (pass)
- `pnpm lint` (0 errors)

File: packages/core/src/feedback-tools.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+      const reportPath = join(this.reportsDir, name);
+      try {
+        if (!statSync(reportPath).isFile()) continue;
+      } catch {
+        continue;
+      }
+
+      const content = readFileSync(reportPath, "utf-8");
+      const parsed = parseReportFile(content);
+      reports.push(parsed);
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply packages/core/src/feedback-tools.ts]: Addressed in commit 28f2d6a.

Resolution:
- Hardened `FeedbackReportStore.list()` to skip unreadable/corrupt files instead of throwing for the entire listing.
- Wrapped read + parse per-file in a guarded block and continue on failure.
- Added regression test: `skips corrupt report files and returns valid ones`.

Evidence:
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts src/__tests__/metadata.test.ts` (pass)
- `pnpm lint` (0 errors)

File: packages/core/src/feedback-tools.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  input: FeedbackToolInput,
+): FeedbackToolInput {
+  return FEEDBACK_TOOL_CONTRACTS[tool].schema.parse(input);
+}
+
+export function generateFeedbackDedupeKey(
+  tool: FeedbackToolName,
+  input: FeedbackToolInput,
+): string {
+  const canonicalEvidence = [...input.evidence].map(normalizeText).sort();
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply packages/core/src/feedback-tools.ts]: Already addressed in commit 5f320a2.

Resolution:
- Dedupe canonicalization now lowercases evidence before sorting.
- Added regression test for case-only ordering stability.

Evidence:
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts` (pass)
- `pnpm lint` (0 errors)

File: packages/core/src/feedback-tools.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    dedupeKey: DedupeKeySchema.parse(raw["dedupeKey"] ?? ""),
+    ...parsedInput,
+  };
+}
+
+function atomicWriteFileSync(filePath: string, content: string): void {
+  const tmpPath = `${filePath}.tmp.${process.pid}.${Date.now()}`;
+  writeFileSync(tmpPath, content, "utf-8");
+  renameSync(tmpPath, filePath);
+}
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply packages/core/src/feedback-tools.ts]: Addressed in commit 68afa3b.

Resolution:
- Extracted `atomicWriteFileSync` into shared module `packages/core/src/atomic-write.ts`.
- Reused this helper in both `packages/core/src/metadata.ts` and `packages/core/src/feedback-tools.ts` to remove duplicated implementations.

Evidence:
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts src/__tests__/metadata.test.ts` (pass)
- `pnpm --filter @composio/ao-core typecheck` (pass)
- `pnpm lint` (0 errors)

File: packages/core/src/feedback-tools.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    .sort();
+  const canonical = [
+    tool,
+    normalizeText(input.title).toLowerCase(),
+    normalizeText(input.body).toLowerCase(),
+    normalizeText(input.session).toLowerCase(),
+    normalizeText(input.source).toLowerCase(),
+    input.confidence.toFixed(4),
+    ...canonicalEvidence,
+  ].join("|");
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply packages/core/src/feedback-tools.ts]: Addressed in commit 68afa3b.

Resolution:
- Replaced delimiter-joined canonical string with structured JSON canonical payload before hashing in `generateFeedbackDedupeKey`.
- This removes deterministic collisions such as `["a|b"]` vs `["a", "b"]`.
- Added regression test: `does not collide when one evidence item contains pipes`.

Evidence:
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts src/__tests__/metadata.test.ts` (pass)
- `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts src/__tests__/paths.test.ts` (pass)
- `pnpm lint` (0 errors)

File: docs/pr-403-feedback-tools-explainer.html [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+          </ul>
+        </article>
+      </section>
+
+      <p class="footer">
+        Scope intentionally excludes fork bootstrap and report->issue automation (per issue #399 constraints).
+      </p>
+    </main>
+  </body>
+</html>
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply docs/pr-403-feedback-tools-explainer.html]: Addressed ✅ Removed both PR-specific artifact files from the branch in commit 9b7a48d: `docs/pr-403-feedback-tools-explainer.html` and `docs/reviews/ao22-pr403-final-status.md`. These details now live in PR discussion instead of permanent docs.

File: docs/reviews/ao22-pr403-confidence-checklist.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+1. Validation is focused on core storage/contracts; no additional AO web/UI flows were exercised here.
+2. OpenClaw agent runs are prompt-sensitive; independent post-run artifact checks were required to confirm side effects.
+3. Duplicate replay behavior is key-stability + separate records (no auto-suppression), which is expected for current scope.
+
+## Final recommendation
+
+MERGE
+
+Reason: all required validation blocks passed, direct OpenClaw dogfood succeeded with independent filesystem verification, and no new functional regressions were detected in PR #403 scope.
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply docs/reviews/ao22-pr403-confidence-checklist.md]: Addressed. Removed PR-specific artifact files from the repo in commit `ee35ced`:
- docs/reviews/ao22-pr403-confidence-checklist.md
- docs/reviews/ao22-work-openclaw-validation-prompt.html

Validation evidence is now kept in PR comments instead of committed artifact files.

File: docs/reviews/ao22-work-openclaw-validation-prompt.html [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  3) dedupe determinism
+  4) operator smoke
+  5) negative tests
+- Environment prerequisites discovered (keys, binaries, profile assumptions).
+- Residual risk list.
+- Final recommendation: MERGE or HOLD.
+
+Important:
+- If any validation step cannot run, mark it clearly as NOT-POSSIBLE and explain why with exact error output.
+- Prioritize factual evidence over summaries.
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply docs/reviews/ao22-work-openclaw-validation-prompt.html]: Resolved in commit `ee35ced` by deleting both PR-specific artifacts with local paths:
- docs/reviews/ao22-pr403-confidence-checklist.md
- docs/reviews/ao22-work-openclaw-validation-prompt.html

No PR-specific review artifacts remain committed.

**Comment by AgentWrapper** [comment]
Readiness update (post-cleanup):

1) Are all Bugbot comments addressed?
- Yes. All Bugbot threads have been addressed and resolved.
- Latest fix commit: `ee35ced` (removed remaining PR-specific artifact files flagged by Bugbot).

2) Is the PR well tested?
- Yes for this scope.
- Core feedback test suite passes: `pnpm --filter @composio/ao-core test -- src/__tests__/feedback-tools.test.ts` (10/10).
- Includes schema validation, malformed confidence, required-field validation, persistence/list behavior, malformed report-file handling, and dedupe determinism.

3) Is test coverage good?
- Good for the implemented scope (#399): contracts, storage behavior, and dedupe-key stability edge cases are covered.
- This PR intentionally does not include fork lifecycle / issue-spawning automation paths.

4) Is the implementation as simple as needed?
- Yes. Minimal explicit schema + strict validation, structured storage reusing existing persistence patterns, and deterministic dedupe key generation.
- No ad-hoc JSON blobs, no out-of-scope automation added.

5) Should we merge?
- Yes, once the currently running CI checks on latest commit complete green.

Current CI state on latest commit:
- Completed green: Lint, Security checks, Web test.
- In progress: Typecheck, Test, Integration Tests, Onboarding Test, Cursor Bugbot.

File: docs/reviews/ao22-design-iteration-summary.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+- Added a dedicated `Consent Gates (Default Policy)` section with hard defaults for non-AO dogfooding:
+  - explicit approval required for `createFork`, `createPR`, and upstream/fork target switching.
+- Clarified that project-level override is optional and only active when explicitly enabled by the owner.
+- Expanded journal behavior in plain language to describe how records are written/updated/retried.
+- Added a minimal journal schema example for the feedback pipeline.
+
+## Ready for review
+
+The docs now provide a deterministic contract for implementing fork-aware report->issue->session->PR execution in follow-up PRs.
```
[the remark this answers is not in the document]
Comment (AgentWrapper) [reply docs/reviews/ao22-design-iteration-summary.md:33]: Good catch. This was missed in PR #403 after merge. Addressed in follow-up PR #420:
- https://github.com/ComposioHQ/agent-orchestrator/pull/420
- Commit: 2229cd9
- Change: removed docs/reviews/ao22-design-iteration-summary.md (remaining PR-specific review artifact).
